File: kernel/processes/credentials.lua
---@class Credentials
---@field ruid number  real user id
---@field euid number  effective user id
---@field suid number  saved set-user-id
---@field rgid number  real group id
---@field egid number  effective group id
---@field sgid number  saved set-group-id
---@field supplementary_groups number[]
local Credentials = {}

function Credentials:new()
	local cred = {
		ruid                 = 0,
		euid                 = 0,
		suid                 = 0,
		rgid                 = 0,
		egid                 = 0,
		sgid                 = 0,
		supplementary_groups = {},
	}
	setmetatable(cred, self)
	self.__index = self
	return cred
end

function Credentials:setuid(uid)
	if self.euid == 0 then
		self.ruid = uid
		self.euid = uid
		self.suid = uid
	else
		-- Non-root may only set euid to ruid or suid; suid is never modified.
		if uid == self.ruid or uid == self.suid then
			self.euid = uid
		end
	end
end

function Credentials:seteuid(uid)
	if self.euid == 0 then
		self.euid = uid
	else
		if uid == self.ruid or uid == self.euid or uid == self.suid then
			self.euid = uid
		end
	end
end

function Credentials:setgid(gid)
	if self.euid == 0 then
		self.rgid = gid
		self.egid = gid
		self.sgid = gid
	else
		if gid == self.rgid or gid == self.egid or gid == self.sgid then
			self.egid = gid
		end
	end
end

function Credentials:setegid(gid)
	if self.euid == 0 then
		self.egid = gid
	else
		if gid == self.rgid or gid == self.egid or gid == self.sgid then
			self.egid = gid
		end
	end
end

-- Returns the effective gid plus any supplementary groups.
function Credentials:get_gids()
	local gids = { self.egid }
	for _, gid in ipairs(self.supplementary_groups) do
		table.insert(gids, gid)
	end
	return gids
end

return Credentials

File: kernel/processes/credentials.ts
export class Credentials {
	ruid: integer = 0;
	euid: integer = 0;
	suid: integer = 0;
	fsuid: integer = 0;
	rgid: integer = 0;
	egid: integer = 0;
	sgid: integer = 0;
	fsgid: integer = 0;
	supplementary_groups: integer[] = [];

	private matches_any_uid(uid: integer) {
		return (uid === this.ruid || uid === this.euid || uid === this.suid);
	}

	private matches_any_gid(gid: integer) {
		return (gid === this.rgid || gid === this.egid || gid === this.sgid);
	}

	private has_perms() {
		// also check for CAP_SETUID?
		return this.euid === 0
	}

	setuid(uid: integer) {
		if (this.has_perms()) {
			this.ruid = this.suid = this.euid = uid;
		} else if (uid === this.ruid || uid === this.suid) {
			this.euid === uid;
		}
	}

	seteuid(uid: integer) {
		if (this.has_perms() || uid === this.ruid || uid === this.suid) {
			this.euid = uid;
		}
	}

	setreuid(ruid: integer, euid: integer) {	}

	setresuid(ruid: integer, euid: integer, suid: integer) {	}

	setgid(gid: integer) {
		if (this.has_perms()) {
			this.rgid = this.egid = this.sgid = gid;
		} else if ( this.matches_any_gid(gid) ) {
			this.egid = gid;
		}
	}

	setegid(gid: integer) {
		if (this.euid === 0) {
			this.egid = gid;
		} else if ( this.matches_any_gid(gid) ) {
			this.egid = gid;
		}
	}

	setregid(rgid: integer, egid: integer) { }

	setresgid(rgid: integer, egid: integer, sgid: integer) { }

	get_gids(): integer[] {
		let gids: integer[] = [ this.egid ];
		for (const gid of this.supplementary_groups) {
			gids.push(gid);
		}
		return gids;
	}
}

export function create(): Credentials {
    return new Credentials();
}
